protocols/rendezvous/src/client.rs

```diff
@@ -314,73 +335,96 @@ impl NetworkBehaviour for Behaviour {
 }
 
 impl Behaviour {
-    fn handle_response(&mut self, request_id: &RequestId, response: Message) -> Event {
+    fn event_for_outbound_failure(&mut self, req_id: &RequestId) -> Option<Event> {
+        if let Some((rendezvous_node, namespace)) = self.waiting_for_register.remove(req_id) {
+            return Some(Event::RegisterFailed(RegisterError::Remote {
+                rendezvous_node,
+                namespace,
+                error: ErrorCode::Unavailable,
+            }));
+        };
+
+        if let Some((rendezvous_node, namespace)) = self.waiting_for_discovery.remove(req_id) {
+            return Some(Event::DiscoverFailed {
+                rendezvous_node,
+                namespace,
+                error: ErrorCode::Unavailable,
+            });
+        };
+
+        None
+    }
+
+    fn handle_response(&mut self, request_id: &RequestId, response: Message) -> Option<Event> {
         match response {
             RegisterResponse(result) => {
-                let (rendezvous_node, namespace) = self
-                    .waiting_for_register
-                    .remove(request_id)
-                    .unwrap_or_else(|| panic!("unknown request_id: {request_id}"));
-                match result {
-                    Ok(ttl) => Event::Registered {
-                        rendezvous_node,
-                        ttl,
-                        namespace,
-                    },
-                    Err(error_code) => Event::RegisterFailed(RegisterError::Remote {
-                        rendezvous_node,
-                        namespace,
-                        error: error_code,
-                    }),
+                if let Some((rendezvous_node, namespace)) =
+                    self.waiting_for_register.remove(request_id)
+                {
+                    return Some(match result {
```

Comment:
I'd suggest we merge these two match by saying:

```rust
match response {
	RegisterResponse(Ok(ttl)) => {}
}
```

It duplicates the `self.waiting_for_register.remove` call but I'd prefer the flatter hierarchy over that.